devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,66 @@
+import sympy
+import numpy as np
+
+from cached_property import cached_property
+
+from devito.tools import filter_ordered, flatten
+
+__all__ = ['Coefficients']
+
+
+class Coefficients(object):
+    """
+    Devito class for users to define custom finite difference weights.
+    """
+ 
+    def __init__(self, *args, **kwargs):
+
+        self.check_args(args, kwargs)
+
+        self.data = args
+        self.rules = self.rules()
+        
+    def check_args(self, args, kwargs):
+        for arg in args:
+            if isinstance(arg, tuple):
+                assert isinstance(arg[0], int)
+                assert(arg[1].is_Function)
+                assert(arg[2].is_Dimension)
+                assert isinstance(arg[3], np.ndarray)
+            else:
+                raise NotImplementedError
+        return
+
+    def rules(self):
+        
+        # FIXME: Move this?
+        def generate_subs(d):
+            
+            deriv_order = d[0]
+            function = d[1]
+            dim = d[2]
+            coeffs = d[-1]
```

Comment:
same as above, I (personally) think it's more readable with explicit function args rather that d[0], d[1]...

Reply:
yes, I agree, I don't quite like this

random thought, potentially completely out of context: perhaps this is a symptom you need subclasses?